File: simulation/engine.js
function simulateWorkflow(patients, isOptimized = false) {
    // Parameters (Mean duration in minutes, Std Dev)
    let checkinTimeParams, waitForRoomParams, encounterParams, docParams;

    if (!isOptimized) {
        // Current State: Inefficiencies present
        checkinTimeParams = { mean: 10, std: 3 };
        waitForRoomParams = { mean: 25, std: 10 }; // Bottleneck!
        encounterParams = { mean: 15, std: 5 };
        docParams = { mean: 12, std: 4 }; // Bottleneck!
    } else {
        // Lean State: Improvements applied
        checkinTimeParams = { mean: 2, std: 1 }; // Self-checkin
        waitForRoomParams = { mean: 5, std: 2 }; // Better flow
        encounterParams = { mean: 15, std: 5 };
        docParams = { mean: 5, std: 2 }; // Templates/Scribes
    }

    // Helper for box-muller transform (Normal distribution)
    function randomNormal(mean, std) {
        let u = 0, v = 0;
        while (u === 0) u = Math.random();
        while (v === 0) v = Math.random();
        let num = Math.sqrt(-2.0 * Math.log(u)) * Math.cos(2.0 * Math.PI * v);
        return (num * std) + mean;
    }

    const results = patients.map(patient => {
        // 1. Arrival -> Check-in
        const checkinDur = Math.max(1, randomNormal(checkinTimeParams.mean, checkinTimeParams.std));
        const checkinEnd = new Date(patient.arrival_time.getTime() + checkinDur * 60000);

        // 2. Waiting Room -> Encounter Start
        const waitDur = Math.max(0, randomNormal(waitForRoomParams.mean, waitForRoomParams.std));
        const encounterStart = new Date(checkinEnd.getTime() + waitDur * 60000);

        // 3. Encounter
        const encounterDur = Math.max(5, randomNormal(encounterParams.mean, encounterParams.std));
        const encounterEnd = new Date(encounterStart.getTime() + encounterDur * 60000);

        // 4. Documentation
        const docDur = Math.max(2, randomNormal(docParams.mean, docParams.std));
        const docEnd = new Date(encounterEnd.getTime() + docDur * 60000);

        const totalLeadTime = (docEnd - patient.arrival_time) / 60000; // minutes

        return {
            patient_id: patient.patient_id,
            checkin_duration: parseFloat(checkinDur.toFixed(1)),
            wait_duration: parseFloat(waitDur.toFixed(1)),
            encounter_duration: parseFloat(encounterDur.toFixed(1)),
            documentation_duration: parseFloat(docDur.toFixed(1)),
            total_lead_time: parseFloat(totalLeadTime.toFixed(1)),
            value_added_time: parseFloat(encounterDur.toFixed(1)),
            simulation_type: isOptimized ? 'Optimized' : 'Current State'
        };
    });

    return results;
}

module.exports = { simulateWorkflow };

File: simulation/engine.py
import pandas as pd
import numpy as np
from datetime import timedelta

def simulate_workflow(df, is_optimized=False):
    """
    Simulates the workflow: Appointment -> Encounter -> Documentation
    
    If is_optimized=True, we apply Lean improvements:
    - Reduced check-in time (e.g., self-service kiosk)
    - Standardized documentation templates (faster documentation)
    - Better resource allocation (reduced wait times)
    """
    
    # Parameters (Mean duration in minutes, Std Dev)
    # Scenario: Current State vs Optimized State
    if not is_optimized:
        # Current State: Inefficiencies present
        checkin_time_params = (10, 3) 
        wait_for_room_params = (25, 10) # Bottleneck!
        encounter_params = (15, 5)
        doc_params = (12, 4) # Bottleneck!
    else:
        # Lean State: Improvements applied
        checkin_time_params = (2, 1) # Self-checkin
        wait_for_room_params = (5, 2) # Better flow
        encounter_params = (15, 5) # Value-added time usually stays similar
        doc_params = (5, 2) # Templates/Scribes
    
    results = []
    
    # We assign resources to simulate simple queuing if needed, 
    # but for this demo, we'll calculate times based on distributions 
    # to show the tracking logic clearly.
    
    for _, row in df.iterrows():
        # 1. Arrival -> Check-in
        checkin_dur = max(1, np.random.normal(*checkin_time_params))
        checkin_end = row['arrival_time'] + timedelta(minutes=checkin_dur)
        
        # 2. Waiting Room -> Encounter Start
        # In a real event sim, this would depend on previous patient. 
        # Here we add random wait plus a cumulative drift if it's busy, 
        # but let's keep it simple: just the "process time" concept.
        wait_dur = max(0, np.random.normal(*wait_for_room_params))
        encounter_start = checkin_end + timedelta(minutes=wait_dur)
        
        # 3. Encounter
        encounter_dur = max(5, np.random.normal(*encounter_params))
        encounter_end = encounter_start + timedelta(minutes=encounter_dur)
        
        # 4. Documentation
        doc_dur = max(2, np.random.normal(*doc_params))
        doc_end = encounter_end + timedelta(minutes=doc_dur)
        
        results.append({
            'patient_id': row['patient_id'],
            'checkin_duration': round(checkin_dur, 1),
            'wait_duration': round(wait_dur, 1),
            'encounter_duration': round(encounter_dur, 1),
            'documentation_duration': round(doc_dur, 1),
            'total_lead_time': round((doc_end - row['arrival_time']).total_seconds() / 60, 1),
            'value_added_time': round(encounter_dur, 1), # Only the consult is "Value Added" in strict Lean terms
            'simulation_type': 'Optimized' if is_optimized else 'Current State'
        })
        
    return pd.DataFrame(results)
